### ISSUE TITLE
Workers take a long time before starting consequent tasks from a single job

### ISSUE THREAD

**Issue by jviotti** [concourse/concourse#1924]
# Bug Report

I'm assuming this is an issue on my pipeline, but I can't seem to be able to find the root cause.

I have a pool of 2 GNU/Linux and 1 macOS worker running a pipeline that builds and tests Electron applications. My pipeline has three tasks:

- `dependencies`
- `test`
- `build-installers`

I noticed that while my tasks run in a reasonable amount of time, the workers seem to be hanging for a while before continuing with the next task. For example, say that the `test` task takes 2 minutes to run, but the worker is stuck for >5 minutes before starting `build-installers`.

Some observations:

- These delays between tasks happen almost on every build
- The delays happen in both GNU/Linux and macOS
- The delay between the `dependencies` and `test` tasks is always way shorter than the delay between the `test` and the `build-installers` tasks
- Sometimes there is a small delay before starting the first task
- This happens even if only one build is running on the corresponding worker

The following can also be handy:

* Concourse version: v3.4.1
* Deployment type (BOSH/Docker/binary): Binary
* Infrastructure/IaaS: Some Intel NUCs and a Mac Mini
* Browser (if applicable): Safari
* Did this used to work? This is my first Concourse setup

This is how my experimental pipeline looks like:

```yaml
# This is how we "import" the `pull-request` third party resource.
# See https://github.com/jtarchie/github-pullrequest-resource
resource_types:
- name: pull-request
  type: docker-image
  source:
    repository: jtarchie/pr
resources:
# These Docker images will be used by the GNU/Linux jobs
# to ensure maximum GLIBC compatibility.
- name: electron-container-x64
  type: docker-image
  source:
    repository: jviottidc/electron-x86_64
    tag: latest
- name: electron-container-x86
  type: docker-image
  source:
    repository: jviottidc/electron-i686
    tag: latest
- name: pull-request
  type: pull-request
  source:
    access_token: ((github-access-token))
    private_key: ((github-ssh-key))
    repo: resin-io/etcher
- name: resin-concourse
  type: git
  source:
    uri: git@github.com:resin-io/resin-concourse.git
    branch: master
    private_key: ((github-ssh-key))
jobs:
- name: tests-darwin-x64
  # The build logs will be publicly accessible
  public: true
  # This option means that only two instances of the job can run at
  # the same time.
  max_in_flight: 2
  plan:
  - get: app-git
    resource: pull-request
    # This means that a change to the pull request resource can
    # trigger a build for this job.
    trigger: true
    version: every
  # This causes GitHub to display a yellow "pending" commit status
  # as soon as the build starts, otherwise we will only see "success"
  # or "failure" once the whole job completes.
  - put: pull-request
    params:
      path: app-git
      context: electron/macos
      status: pending
  - get: electron-container-x64
  - get: resin-concourse
  - task: dependencies
    image: electron-container-x64
    config:
      platform: darwin
      inputs:
        - name: app-git
        - name: resin-concourse
      outputs:
        - name: app-ready
      # TODO: This shouldn't be necessary. Investigate.
      params:
        npm_config_unsafe_perm: true
      run:
        path: /bin/sh
        args:
          - -exc
          - |
            # We want to preserve the dependencies installed on this
            # step on the subsequent tasks. For this reason, we declare
            # a task output called `app-ready`, copy the whole app code
            # there, and install dependencies there. 
            # Then, every other task can require `app-ready` as input,
            # which will have the dependencies inside.
            cp -a app-git/. app-ready
            ./resin-concourse/pipelines/electron/scripts/system-dependencies.sh \
              -b app-ready
            ./resin-concourse/pipelines/electron/scripts/npm-install.sh \
              -b app-ready \
              -r x64 \
              -t electron \
              -s darwin
            # Concourse will `cp` inputs to any subsequent tasks. The 
            # `node_modules/.bin` directory contains symlinks that get 
            # resolved by a recursive copy operation meaning that they 
            # will not work as expected. In order to workaround this, we 
            # create a tarball out of the `.bin` directory (which will 
            # preserve the file attributes) and store it inside the 
            # output directory of this task.
            # Subsequent tasks will need to decompress this tarball 
            # before running any command.
            tar czf app-ready/node_modules/bin.tar.gz app-ready/node_modules/.bin
    params:
      TARGET_ARCH: x64
  - task: test
    image: electron-container-x64
    config:
      platform: darwin
      inputs:
        - name: app-ready
        - name: resin-concourse
      params:
        npm_config_unsafe_perm: true
      run:
        path: /bin/sh
        args:
          - -exc
          - |
            tar fx ./app-ready/node_modules/bin.tar.gz
            rm ./app-ready/node_modules/bin.tar.gz
            ./resin-concourse/pipelines/electron/scripts/system-dependencies.sh \
              -b app-ready
            ( cd app-ready && npm run concourse-test )
            ./resin-concourse/scripts/shared/ensure-npm-valid-dependencies.sh -b app-ready
            ./resin-concourse/scripts/shared/ensure-staged-shrinkwrap.sh -b app-ready
    params:
      TARGET_ARCH: x64
  - task: build-installers
    image: electron-container-x64
    config:
      platform: darwin
      inputs:
        - name: app-ready
        - name: resin-concourse
      # TODO: This shouldn't be necessary. Investigate.
      params:
        npm_config_unsafe_perm: true
      run:
        path: /bin/sh
        args:
          - -exc
          - |
            tar fx ./app-ready/node_modules/bin.tar.gz
            rm ./app-ready/node_modules/bin.tar.gz
            ./resin-concourse/pipelines/electron/scripts/system-dependencies.sh \
              -b app-ready
            ( cd app-ready && npm run concourse-build-installers )
    params:
      TARGET_ARCH: x64
  # Set GitHub commit status accordingly
  on_success:
    put: pull-request
    params:
      path: app-git
      context: electron/macos
      status: success
  on_failure:
    put: pull-request
    params:
      path: app-git
      context: electron/macos
      status: failure
- name: tests-linux-x64
  # The build logs will be publicly accessible
  public: true
  # This option means that only two instances of the job can run at
  # the same time.
  max_in_flight: 2
  plan:
  - get: app-git
    resource: pull-request
    # This means that a change to the pull request resource can
    # trigger a build for this job.
    trigger: true
    version: every
  # This causes GitHub to display a yellow "pending" commit status
  # as soon as the build starts, otherwise we will only see "success"
  # or "failure" once the whole job completes.
  - put: pull-request
    params:
      path: app-git
      context: electron/linux-x64
      status: pending
  - get: electron-container-x64
  - get: resin-concourse
  - task: dependencies
    image: electron-container-x64
    config:
      platform: linux
      inputs:
        - name: app-git
        - name: resin-concourse
      outputs:
        - name: app-ready
      run:
        path: /bin/sh
        args:
          - -exc
          - |
            # We want to preserve the dependencies installed on this
            # step on the subsequent tasks. For this reason, we declare
            # a task output called `app-ready`, copy the whole app code
            # there, and install dependencies there. 
            # Then, every other task can require `app-ready` as input,
            # which will have the dependencies inside.
            cp -a app-git/. app-ready
            ./resin-concourse/pipelines/electron/scripts/system-dependencies.sh \
              -b app-ready
            ./resin-concourse/pipelines/electron/scripts/npm-install.sh \
              -b app-ready \
              -r x64 \
              -t electron \
              -s linux
            # Concourse will `cp` inputs to any subsequent tasks. The 
            # `node_modules/.bin` directory contains symlinks that get 
            # resolved by a recursive copy operation meaning that they 
            # will not work as expected. In order to workaround this, we 
            # create a tarball out of the `.bin` directory (which will 
            # preserve the file attributes) and store it inside the 
            # output directory of this task.
            # Subsequent tasks will need to decompress this tarball 
            # before running any command.
            tar czf app-ready/node_modules/bin.tar.gz app-ready/node_modules/.bin
    params:
      TARGET_ARCH: x64
  - task: test
    image: electron-container-x64
    config:
      platform: linux
      inputs:
        - name: app-ready
        - name: resin-concourse
      params:
        npm_config_unsafe_perm: true
      run:
        path: /bin/sh
        args:
          - -exc
          - |
            tar fx ./app-ready/node_modules/bin.tar.gz
            rm ./app-ready/node_modules/bin.tar.gz
            ./resin-concourse/pipelines/electron/scripts/system-dependencies.sh \
              -b app-ready
            # Electron test commands usually involve `electron-mocha` or 
            # Spectron, which need to execute an actual Electron instance 
            # to perform their job.
            # That won't run in a headless GNU/Linux worker unless we
            # run the command with `xvfb-run`.
            ( cd app-ready && xvfb-run --server-args=$XVFB_ARGS npm run concourse-test )
            ./resin-concourse/scripts/shared/ensure-npm-valid-dependencies.sh -b app-ready
            ./resin-concourse/scripts/shared/ensure-staged-shrinkwrap.sh -b app-ready
    params:
      TARGET_ARCH: x64
  - task: build-installers
    image: electron-container-x64
    # Docker needs to run in privileged node in order to use FUSE,
    # which is required to build AppImages.
    privileged: true
    config:
      platform: linux
      inputs:
        - name: app-ready
        - name: resin-concourse
      run:
        path: /bin/sh
        args:
          - -exc
          - |
            tar fx ./app-ready/node_modules/bin.tar.gz
            rm ./app-ready/node_modules/bin.tar.gz
            ./resin-concourse/pipelines/electron/scripts/system-dependencies.sh \
              -b app-ready
            ( cd app-ready && npm run concourse-build-installers )
    params:
      TARGET_ARCH: x64
  # Set GitHub commit status accordingly
  on_success:
    put: pull-request
    params:
      path: app-git
      context: electron/linux-x64
      status: success
  on_failure:
    put: pull-request
    params:
      path: app-git
      context: electron/linux-x64
      status: failure
- name: tests-linux-x86
  # The build logs will be publicly accessible
  public: true
  # This option means that only two instances of the job can run at
  # the same time.
  max_in_flight: 2
  plan:
  - get: app-git
    resource: pull-request
    # This means that a change to the pull request resource can
    # trigger a build for this job.
    trigger: true
    version: every
  # This causes GitHub to display a yellow "pending" commit status
  # as soon as the build starts, otherwise we will only see "success"
  # or "failure" once the whole job completes.
  - put: pull-request
    params:
      path: app-git
      context: electron/linux-x86
      status: pending
  - get: electron-container-x86
  - get: resin-concourse
  - task: dependencies
    image: electron-container-x86
    config:
      platform: linux
      inputs:
        - name: app-git
        - name: resin-concourse
      outputs:
        - name: app-ready
      run:
        path: /bin/sh
        args:
          - -exc
          - |
            # We want to preserve the dependencies installed on this
            # step on the subsequent tasks. For this reason, we declare
            # a task output called `app-ready`, copy the whole app code
            # there, and install dependencies there. 
            # Then, every other task can require `app-ready` as input,
            # which will have the dependencies inside.
            cp -a app-git/. app-ready
            ./resin-concourse/pipelines/electron/scripts/system-dependencies.sh \
              -b app-ready
            ./resin-concourse/pipelines/electron/scripts/npm-install.sh \
              -b app-ready \
              -r x86 \
              -t electron \
              -s linux
            # Concourse will `cp` inputs to any subsequent tasks. The 
            # `node_modules/.bin` directory contains symlinks that get 
            # resolved by a recursive copy operation meaning that they 
            # will not work as expected. In order to workaround this, we 
            # create a tarball out of the `.bin` directory (which will 
            # preserve the file attributes) and store it inside the 
            # output directory of this task.
            # Subsequent tasks will need to decompress this tarball 
            # before running any command.
            tar czf app-ready/node_modules/bin.tar.gz app-ready/node_modules/.bin
    params:
      TARGET_ARCH: x86
  - task: test
    image: electron-container-x86
    config:
      platform: linux
      inputs:
        - name: app-ready
        - name: resin-concourse
      params:
        npm_config_unsafe_perm: true
      run:
        path: /bin/sh
        args:
          - -exc
          - |
            tar fx ./app-ready/node_modules/bin.tar.gz
            rm ./app-ready/node_modules/bin.tar.gz
            ./resin-concourse/pipelines/electron/scripts/system-dependencies.sh \
              -b app-ready
            # Electron test commands usually involve `electron-mocha` or 
            # Spectron, which need to execute an actual Electron instance 
            # to perform their job.
            # That won't run in a headless GNU/Linux worker unless we
            # run the command with `xvfb-run`.
            ( cd app-ready && xvfb-run --server-args=$XVFB_ARGS npm run concourse-test )
            ./resin-concourse/scripts/shared/ensure-npm-valid-dependencies.sh -b app-ready
            ./resin-concourse/scripts/shared/ensure-staged-shrinkwrap.sh -b app-ready
    params:
      TARGET_ARCH: x86
  - task: build-installers
    image: electron-container-x86
    # Docker needs to run in privileged node in order to use FUSE,
    # which is required to build AppImages.
    privileged: true
    config:
      platform: linux
      inputs:
        - name: app-ready
        - name: resin-concourse
      run:
        path: /bin/sh
        args:
          - -exc
          - |
            tar fx ./app-ready/node_modules/bin.tar.gz
            rm ./app-ready/node_modules/bin.tar.gz
            ./resin-concourse/pipelines/electron/scripts/system-dependencies.sh \
              -b app-ready
            ( cd app-ready && npm run concourse-build-installers )
    params:
      TARGET_ARCH: x86
  # Set GitHub commit status accordingly
  on_success:
    put: pull-request
    params:
      path: app-git
      context: electron/linux-x86
      status: success
  on_failure:
    put: pull-request
    params:
      path: app-git
      context: electron/linux-x86
      status: failure
```

**Comment by jviotti** [comment]
I'm not sure if this is related, but I see the following on the macOS worker node while my pipeline is stuck:

```json
{"timestamp":"1514470359.236778021","source":"baggageclaim","message":"baggageclaim.repository.get-volume.volume-not-found","log_level":1,"data":{"session":"1.8218","volume":"2eb8c98a-db6e-4d1e-592a-c8dcf442bf7d"}}
{"timestamp":"1514470359.236831903","source":"baggageclaim","message":"baggageclaim.api.volume-server.get-volume.volume-not-found","log_level":1,"data":{"session":"2.1.395","volume":"2eb8c98a-db6e-4d1e-592a-c8dcf442bf7d"}}
{"timestamp":"1514470359.782094002","source":"baggageclaim","message":"baggageclaim.repository.get-volume.volume-not-found","log_level":1,"data":{"session":"1.8220","volume":"a85dc35c-46e2-469e-7abc-184ea8cc737b"}}
{"timestamp":"1514470359.782161951","source":"baggageclaim","message":"baggageclaim.api.volume-server.get-volume.volume-not-found","log_level":1,"data":{"session":"2.1.400","volume":"a85dc35c-46e2-469e-7abc-184ea8cc737b"}}
{"timestamp":"1514470360.096256971","source":"baggageclaim","message":"baggageclaim.repository.get-volume.volume-not-found","log_level":1,"data":{"session":"1.8223","volume":"abedf38a-f6fc-4335-5666-1c1b858acfa6"}}
{"timestamp":"1514470360.096292019","source":"baggageclaim","message":"baggageclaim.api.volume-server.get-volume.volume-not-found","log_level":1,"data":{"session":"2.1.406","volume":"abedf38a-f6fc-4335-5666-1c1b858acfa6"}}
{"timestamp":"1514470372.308161259","source":"tsa","message":"tsa.connection.channel.forward-worker.heartbeat.start","log_level":1,"data":{"remote":"10.10.0.104:55438","session":"89.1.1.84","worker-address":"127.0.0.1:36647","worker-platform":"darwin","worker-tags":""}}
{"timestamp":"1514470372.318094492","source":"tsa","message":"tsa.connection.channel.forward-worker.heartbeat.reached-worker","log_level":0,"data":{"baggageclaim-took":"7.54358ms","garden-took":"2.262516ms","remote":"10.10.0.104:55438","session":"89.1.1.84"}}
{"timestamp":"1514470372.326009512","source":"tsa","message":"tsa.connection.channel.forward-worker.heartbeat.done","log_level":1,"data":{"remote":"10.10.0.104:55438","session":"89.1.1.84","worker-address":"127.0.0.1:36647","worker-platform":"darwin","worker-tags":""}}
{"timestamp":"1514470384.470192194","source":"worker","message":"worker.garden.garden-server.destroy.failed","log_level":2,"data":{"error":"unknown handle: 5c8b8ba2-e0d4-4fac-48d4-116a62ea9063","handle":"5c8b8ba2-e0d4-4fac-48d4-116a62ea9063","session":"2.1.600"}}
{"timestamp":"1514470384.470332861","source":"worker","message":"worker.garden.garden-server.destroy.failed","log_level":2,"data":{"error":"unknown handle: 9f88e555-8ddb-44f5-5daf-e1421240b1c0","handle":"9f88e555-8ddb-44f5-5daf-e1421240b1c0","session":"2.1.604"}}
{"timestamp":"1514470384.470397949","source":"worker","message":"worker.garden.garden-server.destroy.failed","log_level":2,"data":{"error":"unknown handle: 6f751c7f-d98a-42b8-785a-0a986576f617","handle":"6f751c7f-d98a-42b8-785a-0a986576f617","session":"2.1.605"}}
{"timestamp":"1514470384.470216036","source":"worker","message":"worker.garden.garden-server.destroy.failed","log_level":2,"data":{"error":"unknown handle: 09054797-3d4a-4bbb-7118-3dfc950e4570","handle":"09054797-3d4a-4bbb-7118-3dfc950e4570","session":"2.1.602"}}
{"timestamp":"1514470384.470227003","source":"worker","message":"worker.garden.garden-server.destroy.failed","log_level":2,"data":{"error":"unknown handle: 538bc367-7e74-466e-70c3-ffe8ecbc6a6f","handle":"538bc367-7e74-466e-70c3-ffe8ecbc6a6f","session":"2.1.603"}}
{"timestamp":"1514470384.470196009","source":"worker","message":"worker.garden.garden-server.destroy.failed","log_level":2,"data":{"error":"unknown handle: 118d561d-eca2-45ee-46ae-ac13091f6e44","handle":"118d561d-eca2-45ee-46ae-ac13091f6e44","session":"2.1.601"}}
{"timestamp":"1514470402.326205492","source":"tsa","message":"tsa.connection.channel.forward-worker.heartbeat.start","log_level":1,"data":{"remote":"10.10.0.104:55438","session":"89.1.1.85","worker-address":"127.0.0.1:36647","worker-platform":"darwin","worker-tags":""}}
{"timestamp":"1514470402.334852457","source":"tsa","message":"tsa.connection.channel.forward-worker.heartbeat.reached-worker","log_level":0,"data":{"baggageclaim-took":"6.313406ms","garden-took":"2.211312ms","remote":"10.10.0.104:55438","session":"89.1.1.85"}}
```

Watching the job using `fly` reveals that is stuck at `initializing`:

```
+ tar czf app-ready/node_modules/bin.tar.gz app-ready/node_modules/.bin
initializing
```

CPU and mem looks OK:

<img width="1280" alt="screen shot 2017-12-28 at 10 17 22 am" src="https://user-images.githubusercontent.com/2192773/34413220-5760443a-ebb8-11e7-9ef6-4d9eae5f4288.png">

**Comment by maritz** [comment]
Seems like a duplicate of #1404 , take a look at the proposed workarounds in #1642 and see if that helps.

For me the naive driver significantly reduced time between steps.

**Comment by vito** [comment]
Consolidating into https://github.com/concourse/concourse/issues/1966 - #1404 is also the overarching issue.
